content/blog/when-to-use-azure-cosmos-db/index.md

```diff
@@ -0,0 +1,290 @@
+
+If you grab our [CosmoDB how to](https://www.pulumi.com/registry/packages/azure-native/how-to-guides/azure-cs-cosmosdb-logicapp/) and `pulumi up` and provision a database, you'll see from that, the first hint of its document roots in that the pulumi namespace is `documentDB`, but it goes deeper because although I can insert and select with sql in Cosmos DB, there is no schema.
+
+### Cosmos DB NoSQL Inserting and Querying
+
+If I insert this customer:
+
+```csharp
+CosmosDatabase database = await client.
+    CreateDatabaseIfNotExistsAsync(databaseId);
+CosmosContainer container = await database.
+    CreateContainerIfNotExistsAsync(containerId, "/customerId");
+
+// Insert a customer document
+var customer = new Customer
+{
+    Id = Guid.NewGuid().ToString(),
+    CustomerId = "12345", // Cosmos DB Partition key
+    Name = "John Doe",
+    Email = "john.doe@example.com"
+};
+
+await InsertCustomer(container, customer);
+```
+
+I can do selects against it like `SELECT * FROM c WHERE c.customerId = '12345"` but when I look in that container in Azure Data Explorer, I see something like this:
+
+```csharp
+{
+    "id": "1d4e26b3-7a02-4e97-8a9d-eccc0b2fd537",
+    "customerId": "12345",
+    "name": "John Doe",
+    "email": "john.doe@example.com",
+    "_rid": "<resource id>",
+    "_self": "<document URI>",
+    "_etag": "\"00000000-0000-0000-0000-000000000000\"",
+    "_attachments": "attachments/",
+    "_ts": 1609459200
+}
+```
+
+It's a JSON document. That's all good, but because there is no schema, as with any document database, if I need to migrate `name` to `fullname` how do I do it? Well, there are a bunch of strategies, but they all involve writing some application code to record by record and move the value from `name` to `full name`.
+
+So don't be fooled by the SQL usage in the API. Cosmos DB is not a relational database; before you use a document database, you should be clear of the trade-offs.
+
+## Document Database / NoSQL Trade-Offs
+
+- **Joins and Relationships**: Cosmos DB lacks support for complex joins and foreign key constraints. Cosmos DB typically requires denormalization or application-level management of relationships.
+
+- **ACID Transactions and Consistency**: Cosmos DB supports ACID transactions but only within single partitions (more on partitions soon). Cross-partition operations in Cosmos DB may require manual consistency handling.
+
+- **Schema Evolution and Migrations**: As shown, relational databases enforce a schema and support migration tools. Cosmos DB's schemaless design offers flexibility but requires custom logic for data migrations, which can increase operational overhead with frequent schema changes.
```

Comment:
"As shown, relational databases enforce a schema..."

Where was this shown?

Reply:
Was in earlier version... but cut because this got too long